File: offer/25.go
/*
 * @Author: lmio
 * @Date: 2023-04-02 14:37:16
 * @LastEditTime: 2023-04-02 14:42:54
 * @FilePath: /leetcode/offer/25.go
 * @Description:剑指 Offer 25. 合并两个排序的链表
 */
package offer

import ."leetcode/code_struct"

func MergeTwoLists(l1, l2 *ListNode) *ListNode {
	dummy := &ListNode{}
	node := dummy
	for l1 != nil && l2 != nil {
		if l1.Val < l2.Val {
			node.Next = l1
			l1 = l1.Next
		}else {
			node.Next = l2
			l2 = l2.Next
		}
		node = node.Next
	}
	if l1 == nil {
		node.Next = l2
	}else {
		node.Next = l1
	}
	return dummy.Next
}

File: code/25.py
'''
Author: lmio 2091319361@qq.com
Date: 2023-07-08 10:20:31
LastEditors: lmio 2091319361@qq.com
Description: 25. K 个一组翻转链表
'''

from typing import Optional

from utils.node import ListNode


def reverseKGroup(head: Optional[ListNode], k: int) -> Optional[ListNode]:
    def reverse(node: ListNode, k: int) -> ListNode:
        prev = node
        if not node.next:
            return None
        cur = prev.next
        # 不断把next插入到prev后面
        i = 1
        while i < k:
            if not cur.next:
                return reverse(prev, i)
            next = cur.next
            cur.next = next.next
            next.next = prev.next
            prev.next = next
            i += 1
        return cur
    node = ListNode()
    node.next = head
    dummy = node
    while node:
        node = reverse(node, k)
    return dummy.next